compiler/rustc_smir/src/rustc_internal/mod.rs

```diff
@@ -24,11 +24,35 @@ use std::ops::Index;
 mod internal;
 pub mod pretty;
 
+/// Convert an internal Rust compiler item into its stable counterpart, if one exists.
+///
+/// # Warning
+///
+/// This function is unstable, and its behavior may change at any point.
+/// E.g.: Items that were previously supported, may no longer be supported, or its translation may
+/// change.
+///
+/// # Panics
+///
+/// This function will panic if StableMIR has not been properly initialized.
 pub fn stable<'tcx, S: Stable<'tcx>>(item: S) -> S::T {
     with_tables(|tables| item.stable(tables))
 }
 
-pub fn internal<'tcx, S: RustcInternal<'tcx>>(item: S) -> S::T {
+/// Convert a stable item into its internal Rust compiler counterpart, if one exists.
+///
+/// # Warning
+///
+/// This function is unstable, and it's behavior may change at any point.
+/// Not every stable item can be converted to an internal one.
+/// Furthermore, items that were previously supported, may no longer be supported in newer versions.
+///
+/// # Panics
+///
+/// This function will panic if StableMIR has not been properly initialized.
+pub fn internal<'tcx, S: RustcInternal<'tcx>>(tcx: TyCtxt<'tcx>, item: S) -> S::T {
+    // The tcx argument ensures that the item won't outlive the type context.
+    let _ = tcx;
```

Comment:
There can be multiple TyCtxt at the same time. You did have to use Lift to ensure that the item comes from the right TyCtxt.

Reply:
can we actually have them at the same time? Or just after each other? (though I guess the same issue applies there)

Reply:
But, that is not an issue here, because once the stable mir context shuts down, there is no more mapping of the stable mir type. You may get the wrong rustc type, but it will still be within the TyCtxt.

Reply:
> can we actually have them at the same time?

Yes, you can a new TyCtxt and enter it and then with access to that TyCtxt it should be possible to create a new stable mir context and pass the TyCtxt you entered yourself as argument to internal(). This would allow the returned value outlive the TyCtxt of the stable mir context despite referencing values that are being freed with the rest of the stable mir context.

Reply:
Ok... I'm amazed our TLS doesn't panic, but just replace the old tls tcx with the new one and switch it back afterwards. I did not expect that.

Reply:
Just to make sure I understand, there could exist 2 or more `TyCtxt` in the same thread, and an item that was interned in one `TyCtxt` might outlive the type in a different one.

In that case, we should invoke `lift` using the `TyCtxt` that was passed to the `intern` function. Is that correct?

It also looks like not everything that implements `RustcInternal`, implements `Lift`, so the intern call will have to be done in the `RustcInternal` implementation for types that require so. 

If both facts are true, my plan is to mark the `RustcInternal` trait as `unsafe` and add `TyCtxt` as an argument to `RustcInternal::internal` as well. It will be up to the `RustcInternal` implementation to invoke `lift`.

Reply:
BTW, this change won't be small. Since the changes in this PR are now unrelated to this issue, can we go ahead and merge the current PR and I can create a follow up PR with the solution I described above. I can even revert adding `tcx` argument to the internal function if you prefer and add it in the next one.